### PR TITLE
log a warning when transactional observers are used

### LINKED ISSUE #3341: Log a warning when transactional observers are used
Transactional observers are not supported yet. We should log a warning when a transactional observer is used.

We should also update the guide and mention the missing feature in the "Limitations" chapter.

### PR DESCRIPTION
**Pull request by machi1990**
Fixes #3341

/cc @mkouba

### CODE REVIEW

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverInfo.java [hunk 1]
```diff
@@ -5,12 +5,14 @@
 import java.util.List;
 import java.util.ListIterator;
 import java.util.Set;
+import java.util.logging.Logger;
```
Comment (gsmet) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverInfo.java]: It should be JBoss Logging not JUL.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverInfo.java [hunk 2]
```diff
@@ -123,12 +125,30 @@ int getPriority() {
...
+                 * Log a warning for transactional observers usage since they are not implemented yet:
+                 * https://github.com/quarkusio/quarkus/issues/2224.
+                 */
+                final AnnotationValue value = annotation.value("during");
+                if (value != null) {
+                    if (AnnotationValue.Kind.ENUM.equals(value.kind())
+                            && value.asEnumType().equals(DotNames.TRANSACTION_PHASE)) {
+                        final ClassInfo clazz = observerMethod.declaringClass();
+                        final String transactionalObserverWarning = String.format(
+                                "The method %s#%s make use of transactional observers which are not implemented yet.", clazz,
```
Comment (gsmet) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverInfo.java]:
```suggestion
                                "The method %s#%s makes use of transactional observers which are not implemented yet.", clazz,
```

**Review by gsmet — CHANGES_REQUESTED** [review]
Thanks. Added a couple of comments.

I let Martin validate it.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverInfo.java (hunk 2 above)
Comment (machi1990) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverInfo.java]: Good catch.

Comment (machi1990) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverInfo.java]: Suggestion incorporated, thanks.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverInfo.java [hunk 3]
```diff
@@ -123,12 +125,28 @@ int getPriority() {
...
                 eventParams.add(annotation.target().asMethodParameter());
+
+                /**
+                 * Log a warning for transactional observers usage since they are not implemented yet:
+                 * https://github.com/quarkusio/quarkus/issues/2224.
+                 */
+                final AnnotationValue value = annotation.value("during");
+                if (value != null) {
+                    if (AnnotationValue.Kind.ENUM.equals(value.kind())
+                            && value.asEnumType().equals(DotNames.TRANSACTION_PHASE)) {
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverInfo.java]: We should only log a warning if `during != TransactionPhase.IN_PROGRESS` which is the default.

**Review by mkouba — CHANGES_REQUESTED** [review]

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverInfo.java (hunk 3 above)
Comment (machi1990) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverInfo.java]: Thanks for the review. I have updated this part and the documentation. WDYT?

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverInfo.java [hunk 4]
```diff
@@ -123,12 +126,27 @@ int getPriority() {
...
             if (Kind.METHOD_PARAMETER == annotation.target().kind()
                     && (annotation.name().equals(DotNames.OBSERVES) || annotation.name().equals(DotNames.OBSERVES_ASYNC))) {
                 eventParams.add(annotation.target().asMethodParameter());
+
+                /**
+                 * Log a warning for usage of transactional observers other than IN_PROGRESS since they are not implemented yet:
+                 * https://github.com/quarkusio/quarkus/issues/2224.
+                 */
+                final AnnotationValue during = annotation.value("during");
+                if (isTransactionalObserver(during)) {
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverInfo.java]: What about `if (isTransactionalObserver(during) && !isDuringTransactionInProgress(during))`?

**Review by mkouba — APPROVED** [review]

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverInfo.java (hunk 3 above)
Comment (mkouba) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverInfo.java]: Looks good.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverInfo.java (hunk 4 above)
Comment (machi1990) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverInfo.java]: I'll push the change shortly.

**Review by gsmet — APPROVED** [review]

**Pull request merged**